lib/ASTGen/Sources/ASTGen/Macros.swift

```diff
@@ -328,34 +461,203 @@ func expandAttachedMacro(
     return 1
   }
 
+  var parentDeclNode: DeclSyntax?
+  if let parentDeclSourceFilePtr = parentDeclSourceFilePtr {
+    parentDeclNode = findSyntaxNodeInSourceFile(
+      sourceFilePtr: parentDeclSourceFilePtr,
+      sourceLocationPtr: parentDeclSourceLocPointer,
+      type: DeclSyntax.self
+    )
+  }
+
+  let customAttrSourceFilePtr = customAttrSourceFilePtr.bindMemory(to: ExportedSourceFile.self, capacity: 1)
+  let declarationSourceFilePtr = declarationSourceFilePtr.bindMemory(to: ExportedSourceFile.self, capacity: 1)
+  let parentDeclSourceFilePtr = parentDeclSourceFilePtr?.bindMemory(to: ExportedSourceFile.self, capacity: 1)
+
+  let discriminatorBuffer = UnsafeBufferPointer(
+    start: discriminatorText, count: discriminatorTextLength
+  )
+  let discriminator = String(decoding: discriminatorBuffer, as: UTF8.self)
+
+  let expandedSources: [String]?
+  switch MacroPluginKind(rawValue: macroKind)! {
```

Comment:
Possibly over engineering here, but we could make a protocol with the two expand functions instead. They could be static or actual methods, where `MacroPluginKind` has an `expander` that returns the type or a new struct (depending on whether you choose static methods or not):
```
enum MacroPluginKind {
...
  var expander: some MacroExpander {
    switch ...
  }
```

Reply:
That's a good idea. Let me do it in a follow-ups because it changes the indentation of the current code, which makes the change set large.